src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.cs

```diff
@@ -184,6 +205,28 @@ public void Reanalyze(IIncrementalAnalyzer analyzer, ReanalyzeScope scope, bool
                 }
             }
 
+            private void OnActiveDocumentChanged(object sender, DocumentId activeDocumentId)
+            {
+                IAsyncToken asyncToken;
+                if (SolutionCrawlerOptions.GetBackgroundAnalysisScope(_registration.Workspace.Options) == BackgroundAnalysisScope.ActiveFile &&
+                    activeDocumentId != null)
+                {
+                    var activeDocument = _registration.Workspace.CurrentSolution.GetDocument(activeDocumentId);
+                    if (activeDocument != null)
+                    {
+                        var lastActiveDocument = Interlocked.Exchange(ref _lastActiveDocument, activeDocument);
+                        if (lastActiveDocument != null)
+                        {
+                            asyncToken = _listener.BeginAsyncOperation("OnDocumentClosed");
+                            EnqueueEvent(lastActiveDocument.Project.Solution, lastActiveDocument.Id, InvocationReasons.DocumentClosed, asyncToken);
+                        }
+
+                        asyncToken = _listener.BeginAsyncOperation("OnDocumentChanged");
+                        EnqueueEvent(activeDocument.Project.Solution, activeDocument.Id, InvocationReasons.DocumentChanged, asyncToken);
```

Comment:
so you can get *multiple* events fired.  i.e. DocumentClosed *and* DocumentChanged will fire, and that's ok/desirable?

Reply:
We are firing DocumentClosed for the previously active document and DocumentChanged for new active document. I switched the latter to now fire DocumentOpened instead of DocumentChanged, which has similar effect, but does not force reanalysis if the incremental analyzer has already computed data for new active document. This seems to have the desired effect for all the existing incremental analyzers, but if we run into problems in future we can consider defining new invocation reason for these and have all the incremental analyzers handle them.